python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,532 @@
+                    job_dag.add_node_if_not_exists(job_container_dict[prereq])
+                    job_dag.add_edge(job_container_dict[prereq], job_container)
+
+                # Cyclic Failure
+                except dag.DAGValidationError:
+                    tester.setStatus('Cyclic or Invalid Dependency Detected!', tester.bucket_fail)
+                    failed_or_skipped_testers.add(tester)
+
+                # Unknown Dependency Failure
+                except KeyError:
+                    tester.setStatus('unknown dependency', tester.bucket_fail)
+                    failed_or_skipped_testers.add(tester)
+
+                # Skipped/Silent/Deleted Testers fall into this catagory, caused by 'job_container' being skipped
+                # during the first iteration above
+                except dag.DAGEdgeDepError:
+                    pass
+
+        # With a working DAG created above (even a partial one), discover race conditions with remaining runnable
+        # testers.
+        failed_or_skipped_testers.update(self.checkRaceConditions(job_dag))
+
+        return failed_or_skipped_testers
+
+    def checkRaceConditions(self, dag_object):
+        """
+        Return a set of failing testers exhibiting race conditions with their
+        output file.
+        """
+        failed_or_skipped_testers = set([])
+
+        # clone the dag so we can operate destructively on the cloned dag
+        dag_clone = dag_object.clone()
+
+        while dag_clone.size():
+            output_files_in_dir = set()
+
+            # Get a list of concurrent tester containers
+            concurrent_jobs = dag_clone.ind_nodes()
+
+            for job_container in concurrent_jobs:
+                tester = job_container.getTester()
+                output_files = tester.getOutputFiles()
+
+                # check if we have colliding output files
+                if len(output_files_in_dir.intersection(set(output_files))):
+
+                    # Fail this concurrent group of testers
+                    for this_job in concurrent_jobs:
+                        tester = this_job.getTester()
+                        tester.setStatus('OUTFILE RACE CONDITION', tester.bucket_fail)
+                        failed_or_skipped_testers.add(tester)
+
+                    # collisions detected, move on to the next set
+                    break
+
+                output_files_in_dir.update(output_files)
+
+            # Delete this group of job containers and allow the loop to continue
+            for job_container in concurrent_jobs:
+                dag_clone.delete_node(job_container)
+
+        return failed_or_skipped_testers
+
+    def schedule(self, testers):
+        """
+        Schedule supplied list of testers for execution.
+        """
+        # If any threads caused an exception, we have closed down the queue and need to
+        # not schedule any more jobs
+        if self.tester_pool._state:
+            return
+
+        job_dag = dag.DAG()
```

Comment:
Put a comment here as to why you are creating the dag outside of the constructor of the TesterData (i.e. shared). Also, consider passing it into the soon to be created "buildDAG" routine.